db/commitlog/commitlog.cc

```diff
@@ -844,65 +867,158 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
+            }
+
+        } catch (...) {
+            ep = std::current_exception();
+            clogger.error("Failed to persist commits to disk for {}: {}", *this, ep);
+        }
+
+        if (flush && !ep) {
+            auto pos = r.file_offset;
+            co_await begin_flush();
+
+            clogger.trace("Flush {} begin at {}", *this, top);
+
+            try {
+                auto prev = pos;
+                if (top < _flushing_to_pos) {
+                    clogger.trace("Flush {} at {} already in progress", *this, top);
+                    co_await _flush_pos.wait(top);
+                } else {
+                    prev = std::exchange(_flushing_to_pos, top);
+                }
+                if (top <= _flush_pos) {
+                    clogger.trace("{} already synced! ({} <= {})", *this, top, _flush_pos);
+                } else {
+                    clogger.trace("Flush {} at {} wait for write {} -> {}", *this, top, _written_pos, top);
+                    co_await _written_pos.wait(top);
+                    clogger.trace("Flush {} at {} wait for previous flush -> {}", *this, prev, pos);
+                    co_await _flush_pos.wait(prev);
+                    co_await _file.flush();
+                    ++_segment_manager->totals.flush_count;
+                    clogger.trace("{} synced to {}", *this, top);
+                    if (_flush_pos < top) {
+                        _flush_pos.set(top);
                     }
-                    // gah, partial write. should always get here with dma chunk sized
-                    // "bytes", but lets make sure...
-                    bytes = align_down(bytes, _alignment);
-                    off += bytes;
-                    view.remove_prefix(bytes);
-                    clogger.trace("Partial write of {} to {}: {}/{} bytes at at {}", bytes, *this, size - view.size_bytes(), size, off - bytes);
-                    continue;
-                    // TODO: retry/ignore/fail/stop - optional behaviour in origin.
-                    // we fast-fail the whole commit.
-                } catch (...) {
-                    clogger.error("Failed to persist commits to disk for {}: {}", *this, std::current_exception());
-                    throw;
                 }
+            } catch (...) {
+                ep = std::current_exception();
+                clogger.error("Failed to flush commits to disk: {}", ep);
             }
-        }, [&]() -> future<> {
-            assert(_pending_ops.has_operation(rp));
-            if (flush_after) {
-                co_await do_flush(top);
+
+            end_flush();
+        }
+
+        // Note: by tradition (and because all counters would go haywire otherwise)
+        // this is updated, exception or no exception.
+        _segment_manager->notify_memory_written(size);
+        _segment_manager->totals.buffer_list_bytes -= buf.size_bytes();
```

Comment:
I don't fully understand this accounting either.
`view` might be smaller than `buf` amd we may have not been able to write all of it.